localstack/extensions/patterns/webapp.py

```diff
@@ -0,0 +1,321 @@
+import importlib
+import logging
+import mimetypes
+import typing as t
+from functools import cached_property
+
+from rolo.gateway import HandlerChain
+from rolo.router import RuleAdapter, WithHost
+from werkzeug.routing import Submount
+
+from localstack import config
+from localstack.aws.api import RequestContext
+from localstack.extensions.api import Extension, http
+
+if t.TYPE_CHECKING:
+    # make sure jinja2 stays optional
+    import jinja2
+
+LOG = logging.getLogger(__name__)
+
+_default = object()
+
+
+class WebAppExtension(Extension):
+    """
+    EXPERIMENTAL! This class is experimental and the API may change without notice.
+
+    A webapp extension serves routes, templates, and static files via a submount and a subdomain through localstack.
+
+    It assumes you have the following directory layout::
+
+        my_extension
+        ├── extension.py
+        ├── __init__.py
+        ├── static              <-- make sure static resources get packaged!
+        │   ├── __init__.py
+        │   ├── favicon.ico
+        │   └── style.css
+        └── templates            <-- jinja2 templates
+            └── index.html
+
+    Given this layout, you can define your extensions in ``my_extension.extension`` like this. Routes defined in the
+    extension itself are automatically registered::
+
+        class MyExtension(WebAppExtension):
+            name = "my-extension"
+
+            @route("/")
+            def index(request: Request) -> Response:
+                # reference `static/style.css` to serve the static file from your package
+                return self.render_template_response("index.html")
+
+            @route("/hello")
+            def hello(request: Request):
+                return {"message": "Hello World!"}
+
+    This will create an extension that localstack serves via:
+
+    * Submount: https://localhost.localstack.cloud:4566/_extension/my-extension
+    * Subdomain: https://my-extension.localhost.localstack.cloud:4566/
```

Comment:
nit / q: If there are a _lot_ of extensions this might become problematic because this could overlap with other domain-based routes. Would it make sense to have them combined in a defined subcomain like `my-extension.extensions.localhost.localstack.cloud`?

Reply:
definitely a good point that we can explore! for now i'll keep this so we don't need to create a new wildcard certificate. for `*.extension....`